examples/generative/finetune_stable_diffusion.py

```diff
@@ -0,0 +1,507 @@
+
+# Load the tokenizer.
+tokenizer = SimpleTokenizer()
+
+#  Method to tokenize and pad the tokens.
+def process_text(caption):
+    tokens = tokenizer.encode(caption)
+    tokens = tokens + [PADDING_TOKEN] * (MAX_PROMPT_LENGTH - len(tokens))
+    return np.array(tokens)
+
+
+# Collate the tokenized captions into an array.
+tokenized_texts = np.empty((len(data_frame), MAX_PROMPT_LENGTH))
+
+all_captions = list(data_frame["caption"].values)
+for i, caption in enumerate(all_captions):
+    tokenized_texts[i] = process_text(caption)
+
+"""
+## Prepare a `tf.data.Dataset`
+
+In this section, we'll prepare a `tf.data.Dataset` object from the input image file paths
+and their corresponding caption tokens. The section will include the following:
+
+* Pre-computation of the text embeddings from the tokenized captions.
+* Loading and augmentation of the input images.
+* Shuffling and batching of the dataset.
+"""
+
+RESOLUTION = 256
+AUTO = tf.data.AUTOTUNE
+POS_IDS = tf.convert_to_tensor([list(range(MAX_PROMPT_LENGTH))], dtype=tf.int32)
+
+augmenter = keras_cv.layers.Augmenter(
+    layers=[
+        keras_cv.layers.CenterCrop(RESOLUTION, RESOLUTION),
+        keras_cv.layers.RandomFlip(),
+        tf.keras.layers.Rescaling(scale=1.0 / 127.5, offset=-1),
+    ]
+)
+text_encoder = TextEncoder(MAX_PROMPT_LENGTH)
+
+
+def process_image(image_path, tokenized_text):
+    image = tf.io.read_file(image_path)
+    image = tf.io.decode_png(image, 3)
+    image = tf.image.resize(image, (RESOLUTION, RESOLUTION))
+    return image, tokenized_text
+
+
+def apply_augmentation(image_batch, token_batch):
+    return augmenter(image_batch), token_batch
+
+
+def run_text_encoder(image_batch, token_batch):
+    return (
+        image_batch,
+        token_batch,
+        text_encoder([token_batch, POS_IDS], training=False),
+    )
+
+
+def prepare_dict(image_batch, token_batch, encoded_text_batch):
+    return {
+        "images": image_batch,
+        "tokens": token_batch,
+        "encoded_text": encoded_text_batch,
+    }
+
+
+def prepare_dataset(image_paths, tokenized_texts, batch_size=1):
+    dataset = tf.data.Dataset.from_tensor_slices((image_paths, tokenized_texts))
+    dataset = dataset.shuffle(batch_size * 10)
+    dataset = dataset.map(process_image, num_parallel_calls=AUTO).batch(batch_size)
+    dataset = dataset.map(apply_augmentation, num_parallel_calls=AUTO)
+    dataset = dataset.map(run_text_encoder, num_parallel_calls=AUTO)
+    dataset = dataset.map(prepare_dict, num_parallel_calls=AUTO)
+    return dataset.prefetch(AUTO)
+
+
+"""
+The baseline Stable Diffusion model was trained using images of 512x512 resolution. It's
+unlikely for a model that's trained using higher resolution images to transfer well to
+lower resolution images. However, the current model will lead to OOM if we keep the
+resolution to 512x512 (without enabling mixed-precision). Therefore, in the interest of
+interactive demonstration, we kept the input resolution to 256x256.
```

Comment:
might be able to workaround this with a batch_size=2 too?

Reply:
Did not get your point.

How does increasing batch size prevent an OOM?

Reply:
Because you're using batch_size=4 here, halving it makes it less likely to get OOM?

Reply:
Oh the current code will OOM out anyway with 512x512 resolution (if not using an A100). Since we're already using 256x256, keeping the batch size to 4 won't hurt IMO.